### ISSUE TITLE
Confused about classification problem with RNN

### ISSUE THREAD

**Issue by Yingyingzhang15** [fchollet/keras#2193]
Hi. I'm confused about the classification problem with RNN.

 I have a  (20000,3)    time sequence and I want to  classify each (1,3) vector in the sequence to its classes(9 classes in total).

According to Keras document, the size of LSTM layer's  input data should be (nb_samples,timesteps,input_dim).

In my case, the nb_samples is 20000,  input_dim is 3. What's the timestep?Is it 1?

Another question. Since the (1,3) vectors are  related to each other in the time domain,  if I set the timestep as 1, can the LSTM layer make use of the time correlation between them?

Thank you for your help!

**Comment by Yingyingzhang15** [comment]
@rpinsler  Can you answer the question for me? Thank you!

**Comment by rpinsler** [comment]
If you want to use RNN, every sample should be a sequence. A timestep of 1 obviously doesn't make sense. What you could do is split your 20000 data points p into equally large samples of t timesteps shifted by 1, e.g.:
x11 = p_{1..t}, y1 = p_t+1
x12 = p_{2..t+1} y2 = p_t+2
...
xn = p_{T-t, T-1}, yn = p_T

To predict the labels for y <= p_t, you have to pad the sequence by adding zeros or something. Then the shape of your dataset would be (n, t, 3).

**Comment by Yingyingzhang15** [comment]
@rpinsler Thank you so much! You helped me a lot!：）

**Comment by Yingyingzhang15** [comment]
@rpinsler Another question, Can I use TimeDistributedDense layer ? It seems that the TimeDistributedDense layer can give every point in every timestep a label.

**Comment by Imorton-zd** [comment]
@rpinsler @rpinsler I don't understand the description.
```
x_t+1 = p_{1..t}, y1 = p_t+1
x_t+2 = p_{2..t+1} y2 = p_t+2
...
xn = p_{T-t, T-1}, yn = p_T
```
Could you give me an example. If one sample is [1,2,3], how can I split the vector into t timesteps shifted? Thanks.

**Comment by Yingyingzhang15** [comment]
I think it's better to describe my question again. I have a time sequence (20000,3) ,where 200000 is the sequence length, and 3 is the input_dim. I want to classify every (1,3) vector to its classes(13 classes in total). 

My code is as follows:
`model = Sequential()`
`model.add(LSTM(150, input_length = sequence_length, input_dim=3, return_sequences=True))`
`model.add(Dropout(dropout))`
`model.add(LSTM(150, return_sequences=True))`
`model.add(Dropout(dropout))`
`model.add(TimeDistributedDense(nb_classes))`
`model.add(Activation('softmax'))`
`model.compile(loss='categorical_crossentropy', optimizer='adam')`

In order to use  LSTM and TimeDistributedDense layer, do I have to split the long (20000,3)  sequence to several short sequences (like (30,3) vector)?

 I want to classify each (1,3) vector according to the context information before and behind this vector.However, if I split the  sequences, in each (30,3) vector, the earlier information is not considered for the (0:10,3) vector labelling. Similarly, the information after the (20:29,3) is not considered.

In a word ,is there any method in Keras with which I can classify the whole long sequence?
Or is there any method with which I can split the long sequence to short sequences as well as considering the context information?

I feel quite confused about this question. Can you help me out? Thank you very much!

@fchollet @EderSantana @rpinsler @farizrahman4u @wxs

**Comment by wxs** [comment]
You don't *have* to split the sequences depending on how large your batches are, but you'll want to use Theano rather than TF since TF unrolls the loop of the RNN so you'll get a mega-graph 20,000 elements long.

However, you're probably better of splitting them, as you say. I think what you're looking for is the `stateful` flag of the RNN. It causes the RNN to retain its internal state between batches, until you call `.reset_states()` (see http://keras.io/layers/recurrent/).

**Comment by Yingyingzhang15** [comment]
@wxs Thank you for your reply! I read the Keras doc about the stateful flag of RNN. And there is another question like the one discussed in #1870 .

Take a simple sequence X=[0,1,2,3,4,...,100] for example.

I want to give every point a label,like:
[0,1,2,3,4]->give the point X[0]=0 label A
[1,2,3,4,5]->give the point X[1]=1 label B
[2,3,4,5,6]->give the point X[2]=2 label C

How should I process the input data？The input of LSTM layer is

>  (nb_sample, timesteps, input_dim)

.So should I set the timesteps as 1, the batch_size as 5?

 However , if I set the batch_size as 5, the first batch is[0,1,2,3,4], and the second batch is [5,6,7,8,9], so the state of the first batch  is retained. But the 0's state is retained to 5, 1's state is retained to 6, which is not my purpose.

What should I do then? I really don't know how to set the `timesteps` in the input of LSTM layer.
I'm looking forward for your answer! Thank you  in advance!

**Comment by wxs** [comment]
If you only had that *one* sequence you'd make your batch_size 1, you would make your "timesteps" a number of your choosing. Different values would affect how often it made gradient-descent weight updates. Let's say you made your timesteps 10: your first training example would look like this:

```
X = 
[[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]]
y =
[[X, X, X, X, A, B, C, D, E]]
```
then it would do a gradient update. The next training example would be

```
X = 
[[10, 11, 12, 13, 14, 15, 16, 17, 18, 19]]
y =
[[F,  G,  H,  I,  J,  K,  L,  M,  N  O,  P]]
```

(you would of course want to make sure to have `stateful=True`). By the time it completed a full epoch it would have done 10 gradient updates (because you had 100 timesteps and you were processing it in sets of 10).

This is a pretty contrived example, though. Normally, you would have multiple training sequences, so your batch_size would be set to a larger number so as to be simultaneously training on multiple sequences.

**Comment by Yingyingzhang15** [comment]
@wxs It's so nice of you for your help! 
I have another question. For example, I have 4 sequences: A,B,C,D.Each sequence is spilted to three short sequences:A1,A2,A3;B1,B2,B3;C1,C2,C3;D1,D2,D3.
If I set the batch_size as 4, and I have `stateful=True`. So the order of my data is 
A1,B1,C1,D1,A2,B2,C2,D2,A3,B3,C3,D3?

**Comment by yinruiqing** [comment]
@wxs I've read your comments. It's very helpful. But I found the label `stateful` is just in the RNN Layer, if I want use LSTM. How can I do it.

**Comment by yinruiqing** [comment]
@Yingyingzhang15 I am doing a same task like you. I also want to use TimeDistributedDense layer. But I meet some problems. I mentioned it in another issue #1029 . Have you been success with TimeDistributedDense layer?

**Comment by rpinsler** [comment]
LSTM inherits from the Recurrent base class, therefore also has the stateful property. It's actually mentioned in the docs:

> Abstract base class for recurrent layers. Do not use in a model -- it's not a valid layer! Use its children classes LSTM, GRU and SimpleRNN instead. All recurrent layers (LSTM, GRU, SimpleRNN) also follow the specifications of this class and accept the keyword arguments listed below.

**Comment by yinruiqing** [comment]
@rpinsler Thank you, I just read it.

**Comment by lynnwong11** [comment]
@Yingyingzhang15 i think batch_size has no direct relationship to timestteps. As in your example, timesteps should be 5, and batch_size should depend on your computer computation ability and is needed to be adjusted.
